### DISCUSSION TITLE
Support for pmtiles

### DISCUSSION THREAD

**Discussion by aemonm** [visgl/deck.gl#7861] [Ideas]
Any plans to support [pmtiles](https://protomaps.com/docs/pmtiles/) in a future release? I believe they are similar to MVTs but are queryable via HTTP range headers rather than the `some-tile-server.com/z/x/y` method.

**Comment by ibgreen** [comment]
- This would be more of a question for loaders.gl. deck.gl leverages loaders.gl for data loading and parsing.
- We don't currently have any plans for pmtiles, but it fits well into the general loaders.gl roadmap. 
- Generally, Use of HTTP range requests has become very fashionable lately and there are a range of new formats adopting them (COGs, COPC, FlatGeobuf, etc) which we'd all love to provide support for at some point.

**Reply by arthurgailes** [reply]
It appears to be added on [loaders](https://loaders.gl/docs/modules/pmtiles/formats/pmtiles); does that place it on deck.gl's timeline?

**Reply by aemonm** [reply]
@ibgreen it'd be great to get an official PMTiles Layer now that support exists in loaders.

**Reply by cboettig** [reply]
@ibgreen Just joining the chorus here: it's great to see that PMTiles is supported in loaders.gl already now, and it would be lovely to be able to access that using deck.gl

**Reply by ibgreen** [reply]
The `TileSourceLayer` that powers the PMTiles demo in loaders.gl is available in the `@deck.gl-community/layers` module.

**Reply by cboettig** [reply]
@ibgreen really nice, thanks!  any pointers for getting `TileSourceLayer` to render in the corresponding pydeck code?  I'm not able to find any docs or examples on the pydeck side with this unfortunately.

**Reply by ibgreen** [reply]
Hmm, no sorry, we haven't yet discussed making pydeck bindings for the deck.gl-community modules. I will try to bring it some attention, but It is a fair bit of work so can't guarantee that someone will pick it up.

**Reply by cboettig** [reply]
Thanks, understood.  I tried experimenting based on https://github.com/agressin/pydeck_myTileLayer without success.  I don't suppose there's any roadmap to make pmtiles a core tile layer to sidestep the challenges of extending generic community modules to pydeck?  The performance difference with pmtiles (e.g. rendered with maplibre) is really remarkable.

**Reply by ibgreen** [reply]
Well, the TIleSource system in loaders.gl is still experimental and the idea was to let this take its time, and we'd publish in the community repo in the mean time to let early adopters start using it.

Cleaning up and reviewing the `Source` APIs and the `TIleSourceLayer` is a fair bit of work. 

It could perhaps be added as an experimental, underscore, will-break-in-future-release export. I'll add it to agenda for next bi-weekly community meeting.

**Reply by prusswan** [reply]
I modified MaticoApp's plugin (which is more or less working except it is for an older version of deck gl and missing the iife build needed for pydeck): https://github.com/prusswan/deck.gl-pmtiles/tree/overture

Probably will be superceded by the PMT reader in loaders.gl but I needed something that can work with pydeck quickly

**Comment by sheecegardezi** [comment]
Yep it's possible now: https://github.com/visgl/deck.gl/issues/8615#issuecomment-2460355607

**Reply by cboettig** [reply]
cool! Is this possible directly from pydeck?  (currently I can't seem to get pydeck to read from other formats listed by loaders.gl, e.g. geoparquet).

**Reply by jusolu** [reply]
Nice!
Same here, there's a way to implement with pydeck?
https://deckgl.readthedocs.io/en/latest/gallery/custom_layer.html

**Reply by sheecegardezi** [reply]
I'm a bit unsure about PyDeck. At the moment, I'm focusing on how I can dynamically add or remove a PMTiles layer.

**Reply by prusswan** [reply]
yes, I have forked MaticoApp's plugin for usage with pydeck, there is an example included
